### PR TITLE
Rename constructors of collections

### LINKED ISSUE #484: Rename constructors of collections
Follow-up from #477

In `array`, `ref`, `bytes`, and other collections:
1. Rename `unsafeFresh` to `allocate` and clearly document that it is unsafe
    - perhaps we should decide if we want to have `collection::unsafeFunction` or `collection::unsafe::function` in general for functions acting on the collections that are generally unsafe
2. Rename `fresh` (formerly `fill`) to `$collection_name` or `$Collection_name`, for example `array` or `Array`
    - we also discussed using `new` instead, but that's: a) ambiguous (as now we need the `array::` qualifier to determine that it's a new _array), and b) clashes with the existing `new` keyword

### PR DESCRIPTION
**Pull request by jiribenes**
## Motivation

Resolves #484 

## Description

In `array`, `ref`, `bytes`:
1. Rename `unsafeFresh` to `allocate` and clearly document that it is unsafe
2. Rename `fresh` (formerly `fill`) to `$collection_name` or `$Collection_name`, for example `array` or `Array`
3. Add documentation comments where needed :)

I also restructured `array` a little bit so that the "constructors" are at the top (intended for easier API discovery).

### Questions

1. Are we happy with the naming scheme of `$collection_name` for the "fill"/"fresh" constructor?
    - if so, are there any other data structures we want to support here? (what about `bytes`, for example?)
2. Should unsafe functions be in their own sub-namespace?
2. Is there anything else we want to add in this PR?

### CODE REVIEW

File: libraries/common/array.effekt [hunk 1]
```diff
@@ -4,24 +4,52 @@ import effekt
...
+ */
+extern global def array[T](size: Int, init: T): Array[T] =
+  ml "Array.array (${size}, ${init})"
+  default {
+    val arr = allocate[T](size);
+    each(0, size) { i =>
+      unsafeSet(arr, i, init)
+    };
+    arr
+  }
```
Comment (b-studios) [inline libraries/common/array.effekt:35]: Is this more efficient on ml this way? Or is this even mandatory?

File: libraries/common/array.effekt [hunk 2]
```diff
@@ -128,24 +157,11 @@ def build[T](size: Int) { index: Int => T }: Array[T] = {
...
-  }
-  return arr;
-}
-
 def toList[T](arr: Array[T]): List[T] = {
   var i = arr.size - 1;
   var l = Nil[T]()
   while (i >= 0) {
-    l = Cons(unsafeGet(arr, i), l)
+    l = Cons(arr.unsafeGet(i), l)
```
Comment (b-studios) [inline libraries/common/array.effekt:164]: If I imagine writing `unsafe::get(arr, i)` or `arr.unsafe::get(i)` I am starting to thing that `unsafeGet` is nicer.

File: libraries/common/io/files.effekt [hunk 3]
```diff
@@ -95,7 +95,7 @@ def readFile(path: String): String / { AsyncIO, Exception[IOError] } = {
 
   val readSize = 1048576 // 1MB
   var size = readSize
-  var buffer = bytes::empty(size)
+  var buffer = bytes::allocate(size)
```
Comment (b-studios) [inline libraries/common/io/files.effekt:98]: Should this be `allocate` or `bytes`? There is no other constructor here.

**Comment by b-studios** [comment]
Great! Thanks a lot for this PR! That cleans up a lot of things

File: libraries/common/array.effekt (hunk 2 above)
Comment (jiribenes) [reply libraries/common/array.effekt:164]: Yeah, same, I originally had `arr.unsafe::get(i)`, but I didn't like the aesthetics of it.

File: libraries/common/array.effekt (hunk 1 above)
Comment (jiribenes) [reply libraries/common/array.effekt:35]: I didn't add this code, I just changed the order of functions in that file, sorry. :)

File: libraries/common/io/files.effekt (hunk 3 above)
Comment (jiribenes) [reply libraries/common/io/files.effekt:98]: It's closest to `array::allocate` in that it's unsafe.

Comment (b-studios) [reply libraries/common/io/files.effekt:98]: It's safe, no? It just initializes with 0s ;)

Comment (jiribenes) [reply libraries/common/io/files.effekt]: I thought about this for a bit and you're right, I've renamed it to just `bytes`.

**Review by b-studios — APPROVED** [review]
Great, thanks!

**Pull request merged**